packages/backend/src/api/controllers/SearchController.ts

```diff
@@ -0,0 +1,113 @@
+import { ControllerResult } from './ControllerResult'
+
+export class SearchController {
+  constructor(
+    private stateUpdateRepository: StateUpdateRepository,
+    private userRegistrationEventRepository: UserRegistrationEventRepository
+  ) {}
+
+  async getSearchRedirect(query: string): Promise<ControllerResult> {
+    const parsed = parseSearchQuery(query)
+
+    let response: ControllerResult | undefined
+
+    if (!response && parsed.ethereumAddress) {
+      response = await this.searchForEthereumAddress(parsed.ethereumAddress)
+    }
+
+    if (!response && parsed.stateTreeHash) {
+      response = await this.searchForRootHash(parsed.stateTreeHash)
+    }
+
+    if (!response && parsed.starkKey) {
+      response = await this.searchForStarkKey(parsed.starkKey)
+    }
+
+    if (!response) {
+      const content = "Search query couldn't be found"
+      return { type: 'not found', content }
+    }
+
+    return response
+  }
+
+  private async searchForEthereumAddress(
+    ethereumAddr: EthereumAddress
+  ): Promise<ControllerResult | undefined> {
+    const userRegistrationEvent =
+      await this.userRegistrationEventRepository.findByEthereumAddress(
+        ethereumAddr
+      )
+    if (userRegistrationEvent === undefined) {
+      return
+    }
+
+    const positionId =
+      await this.stateUpdateRepository.getPositionIdByPublicKey(
+        userRegistrationEvent.starkKey
+      )
+    if (positionId === undefined) {
+      return
+    }
+
+    return { type: 'redirect', url: `/positions/${positionId}` }
+  }
+
+  private async searchForRootHash(
+    hash: PedersenHash
+  ): Promise<ControllerResult | undefined> {
+    const stateUpdateId =
+      await this.stateUpdateRepository.getStateUpdateIdByRootHash(hash)
+    if (stateUpdateId === undefined) {
+      return
+    }
+
+    return { type: 'redirect', url: `/state-updates/${stateUpdateId}` }
+  }
+
+  private async searchForStarkKey(
+    starkKey: StarkKey
+  ): Promise<ControllerResult | undefined> {
+    const userRegistrationEvent =
+      await this.userRegistrationEventRepository.findByStarkKey(starkKey)
+    if (userRegistrationEvent === undefined) {
+      return
+    }
+
+    const positionId =
+      await this.stateUpdateRepository.getPositionIdByPublicKey(
+        userRegistrationEvent.starkKey
+      )
+    if (positionId === undefined) {
+      return
+    }
+
+    return { type: 'redirect', url: `/positions/${positionId}` }
+  }
+}
+
+interface ParsedQuery {
+  ethereumAddress?: EthereumAddress
+  stateTreeHash?: PedersenHash
+  starkKey?: StarkKey
+}
+
+export function parseSearchQuery(query: string): ParsedQuery {
+  if (query.startsWith('0x') && query.length === 66) {
+    return {
+      stateTreeHash: PedersenHash(query),
+      starkKey: StarkKey(query),
+    }
+  }
+  if (query.startsWith('0x') && query.length === 42) {
+    return {
+      ethereumAddress: EthereumAddress(query),
+    }
+  }
```

Comment:
Because `PedersenHash`, `StarkKey` and `EthereumAddress` can throw and you're partially duplicating their logic here I think it's better to use the following structure:
```ts
const parsed: ParsedQuery = {}
try {
  parsed.stateTreeHash: PedersenHash(query),
} catch {}
try {
  parsed.starkKey: StarkKey(query),
} catch {}
try {
  parsed.ethereumAddress: EthereumAddress(query),
} catch {}
return parsed
```

Reply:
Good idea! I refactored it using `tryOrUndefined`. WDYT? is there a good place for such util?

Reply:
Below is great